aws-datastore/src/main/java/com/amplifyframework/datastore/syncengine/SyncProcessor.java

```diff
@@ -92,12 +106,67 @@ public static Builder builder() {
     Completable hydrate() {
         ModelWithMetadataComparator modelWithMetadataComparator =
             new ModelWithMetadataComparator(modelProvider, modelSchemaRegistry);
-
+        final Map<String, ModelSyncMetrics> metricsByModel = new ConcurrentHashMap<>();
         final Set<Completable> hydrationTasks = new HashSet<>();
         for (Class<? extends Model> clazz : modelProvider.models()) {
+            SyncTime lastSyncTime = syncTimeRegistry.lookupLastSyncTime(clazz)
+                .map(this::filterOutOldSyncTimes)
+                .blockingGet();
+            metricsByModel.put(clazz.getSimpleName(), new ModelSyncMetrics(lastSyncTime));
             hydrationTasks.add(createHydrationTask(modelWithMetadataComparator, clazz));
         }
-        return Completable.concat(hydrationTasks);
+
+        // We'll collect metrics for the initial sync by observing changes to the local DataStore.
+        Cancelable metricsObserver =
+            localStorageAdapter.observe(itemChange -> {
```

Comment:
Since we have the observer pattern here, I don't think you even need to put this in the `SyncProcessor` at all, do you? You could have this be like a `SyncMetricsPublisher.java` or a `SyncMetricsObserver.java`, and then cleanly separate all of this logic into its own little place.

Overall though, I'm worry about how long `hydrate()` is getting, here. Any approaches you can think of to break this apart into some smaller bits would probably useful, towards keeping the code clear. -- Keep in mind, the reader expects `hydrate()` to be _mostly_ about populating data, not broadcasting events.

The other benefit to breaking out a separate observer class is that you will no longer have `SyncProcessor` depending on the `LocalStorageAdapter`. I don't think I ever _completely_ got there, but you can se several mid-level utilities like the `VersionRepository`, etc., which abstract away the `LocalStorageAdapter` from the sync processor. All in all, I think the `LocalStorageAdapter` is a fairly low level detail for the sync processor and -- where possible -- it would be best to avoid direct dependency on it.

Reply:
LOL...I should have followed my gut on this one. I was going to do just that and talked myself out of it. 🤣  The moment I had to add the local `LocalStorageAdapter` as a dependency, it felt wrong. Glad to refactor this though.

Reply:
I refactored this on the latest commit.